### DISCUSSION TITLE
mc-cost-optimizer (v0.3.0) related issues and discussions

### DISCUSSION THREAD

**Discussion by seokho-son** [m-cmp/m-cmp#11] [General]
## mc-cost-optimizer 기술 개발 목표 및 현황

(참고: 개발 진행 현황은 '24 2차 컨소시움 공유 자료)

o **멀티 클라우드 서비스 비용 상태 분석/알람 관리 개발 및 통합 시험**
- 클라우드별 서비스 비용 데이터 수집 기술 개발 (예정)
- 클라우드별 서비스 비용 데이터 아카이빙 기술 개발 (예정)
- 멀티 클라우드 서비스 비용 상태 분석/알람 관리 통합 시험 **(개발완료: Mailing, Slack 등)**

o **멀티 클라우드 서비스 비용 분석/리포팅 관리 개발 및 통합 시험**
- 멀티 클라우드 서비스 비용 정보 필터링 및 비용 추적 기술 개발 **(개발완료)**
- 멀티 클라우드 서비스 비용 분석 정보 가시화 기술 개발 **(개발완료)**
- 멀티 클라우드 서비스 비용 분석/리포팅 관리 통합 시험 **(개발완료)**

o **멀티 클라우드 과금 정보 통합 관리 개발 및 통합 시험**
- 클라우드별 과금 정보 수집 기술 개발 (부분진행)
- 클라우드별 과금 정보 아카이빙 기술 개발 (부분진행)
- 사용/지원/할인 요금, 크레딧 적용 기술 개발 (부분진행)
- 멀티 클라우드 과금 정보 통합 관리 통합 시험 (_**2025년 추진으로 이관 요청**_)

o **멀티 클라우드 서비스 비용 최적화 정책 관리 개발 및 통합 시험**
- 멀티 클라우드 서비스 비용 최적화 정책 규격 보완 (예정)
- 멀티 클라우드 서비스 비용 최적화 정책 관리 기술 개발 (예정)
- 멀티 클라우드 비용 절감을 위한 미사용/유휴 자원 및 서비스 최적화 기술 개발 (예정)
- 멀티 클라우드 서비스 비용 최적화 정책 관리 통합 시험 (예정)

o **멀티 클라우드 비용 관리 최적화 대시보드 설계 및 개발**
- 멀티 클라우드 서비스 비용 상태 분석 및 알람 대시보드 기능 분석 및 화면 설계 (개발중)
- 멀티 클라우드 서비스 비용 상태 분석 및 알람 대시보드 개발 (개발중)
- 멀티 클라우드 과금 정보 통합 관리 대시보드 기능 분석 및 화면 설계 **(개발완료)**
- 멀티 클라우드 과금 정보 통합 관리 대시보드 개발 **(개발완료)**

관련 주요 성능 목표치
- 3종 서비스 운용 최적화 정책 지원 (리드: mc-cost-optimizer)

## 이슈 사항 리스트

- 기본 가이드 및 DockerFile 제공 등 필요 
  - https://github.com/m-cmp/mc-cost-optimizer/issues/8
  - 시급: ~~진행중~~, 1차 개발 완료/오류대응 중

- 3종 서비스 운용 최적화 정책 지원 (리드: mc-cost-optimizer) 달성 방안 확정 필요
  - 현재 제시된 목록 확인 필요 (참고: 최종 목표인 최적화 정책 6개는 cost-optimizer에서 항목 제안 예정 (8월 말))
    - 이상비용탐지
    - 미사용자원탐지
    - Right Sizing 정책    
  - 최적화 구현 방법 확인 필요
    - o11y 모니터링 기능 활용 예정 (현재는 제공하는 모듈이 있다고 가정하게 개발 진행 중)
    - 0.3.0 릴리스에 포함시, o11y 모니터링에 기능 및 일정 의존성 발생

- 지원 CSP 종류 확인 : (v0.3.0) AWS / (‘25) Azure, GCP, NCP
  - 수집 모듈을 자체 개발하므로, 대상 CSP는 제한적으로 지원
  - AWS 수집 모듈 개발 지연 가능성이 있음 
    - 0.3.0 릴리스에서는 목업 데이터로 처리 가능 여부 확인 필요

- 비용 지원 자원 종류 확인 
  - VM 인스턴스
  - K8s 서비스(일정상 지원 가능한 경우에 추진)
  - 기타(?)

- CB-TB, o11y 등과 연동 시, 리소스 운용 정보 활용 방법 등 협의 필요 있음
  - cost-optimizer 는 CSP를 통해 리소스 운용 데이터(cost 등)를 직접 받게되며, M-CMP 플랫폼에서 생성된 리소스와는 매핑 작업이 필요함. 
    - 방법이 확정되어 있지는 않음
      - CB-TB 리소스 운용 정보를 API를 통해 호출하여 획득할 수도 있으며, 
      - web-console을 통해서 직접 입력 받을 수도 있음. 
      - CSP 리소스의 tagging을 통해 매핑하는 것도 방법일 수 있음.
      
- 최적화 정책에 따른 액션 수행 필요성 및 방안
  - cost-optimizer가 어디에 요청을 해야하는지 논의 및 확인 필요
    - Web-Console에 정책만 알림(?) / 직접 최적화 액션 수행(?)
  - 릴리스 대상여부 확인 필요
  
- API 문서 및 기능 제공 필요
  - 대상 기능 확정 필요 (릴리스 기준): TBD
  - 참고: API는 Swagger 로 제공되어야 함. ([스펙은 OpenAPI 2.0으로 제공 요망](https://github.com/m-cmp/m-cmp/issues/2))

**Comment by seokho-son** [comment]
이슈 사항 추가
- 별도의 GUI를 개발 중이며, 해당 GUI는 향후 web-console에 iframe 등으로 embedding 되어야 하므로, 상호 일관성을 높여서 개발할 필요가 있음. (web-console 과 상호 검증 및 협의 필요)

**Comment by seokho-son** [comment]
@BSjaeyoung-kim 혹시 추가적으로 정리가 필요한 이슈가 있으시면, 공유 부탁드립니다. :)

**Comment by BSjaeyoung-kim** [comment]
@seokho-son discussion이 이쪽에도 있는걸 놓치고 있었습니다 ㅎㅎ
현재 docker-compose 부분은 현재 merge 완료한 상황입니다. 
이슈사항으로 적어두신 최적화 정책 제공사항과 연동방안에 대한 내용은 금주 중으로 확인하여 해당 discussion에 내용 업데이트 하겠습니다.

감사합니다

**Comment by seokho-son** [comment]
@BSjaeyoung-kim 안녕하세요! 

비용 수집 및 조회 관련 하여, 리소스별 지원 대상 문의 드립니다. 아래 리소스 범주를 기준으로 공유해주시면 감사하겠습니다. 
(고려하고 계시는, 올해 릴리스 목표 및 최종 목표)

- VM 인스턴스
- K8s 서비스 및 인스턴스
- 기타 (NET, NET traffic, ...)

예를 들면, 
- VM 인스턴스 (올해: 지원, 최종: 지원)
- K8s 서비스 및 인스턴스 (올해: 부분지원, 최종: 지원, 비고: xxxx)
- Net ...
- ...

최종 목표가 불분명하다면, 우선 v0.3.0 릴리스 목표를 기준으로 공유해주셔도 무방합니다. :)

**Reply by BSjaeyoung-kim** [reply]
비용에 대해서는 v0.3.0 릴리스 기준으로 VM인스턴스는 확정으로 적용할 예정입니다.

이외 비용에 대해서는 CSP에서 수집가능한 모든 비용에 대해서는 비용정보를 제공할 수 있으나 비용최적화를 제공할 수 있는 인스턴스에 제약이 있을 것으로 생각됩니다.(비용최적화의 경우도 v0.3.0에서는 VM 인스턴스까지 적용 예정입니다)

또한 최종적으로 제공할 수 있는 최적화 기능에 대해서도, CSP별, 서비스 별 차이가 존재하기 때문에, 해당 논의는 이전에 말씀드린대로 8월말까지 정리하여 공유드릴수 있을것 같습니다.

상세 인스턴스별 비용 출력에 대해 현재 구상하는 (안)에 대한 스크린샷 함께 첨부드립니다.
<img width="280" alt="스크린샷 2024-08-08 오후 4 48 37" src="https://github.com/user-attachments/assets/0a4729fe-54a1-4478-940d-d1ffe2e7eeda">

**Reply by BSjaeyoung-kim** [reply]
- 기본 가이드 및 DockerFile 제공 등 필요
 -- 기본가이드 동영상 형태로 제공 예정
 -- docker compose build 설정 및 테스트 완료
 -- workflow에 대해서는 확인 및 작업 중

- 3종 서비스 운용 최적화 정책 지원 (리드: mc-cost-optimizer) 달성 방안 확정 필요
 -- v0.3.0 release 시 개발 내역, 이상비용탐지(확정), 미사용자원탐지(확정), Right sizing(조금더 고려 필요)

- 최적화 구현 방법 확인 필요
 -- 지원 CSP 종류 확인 : (v0.3.0) AWS / (‘25) Azure, GCP, NCP
 -- 현재 모니터링에 대한 데이터를 일정정도 예상하여 개발 진행중

- 수집 모듈을 자체 개발하므로, 대상 CSP는 제한적으로 지원 
 -- AWS의 경우 실제 수집된 데이터를 dummy처리하여 개발 중이며, 수집모듈 또한 개발 착수 중
 -- 개발일정이 밀릴 수는 있지만, 현재 11월 말까지 개발 완료 목표로 진행중

- 비용 지원 자원 종류 확인
 -- 비용의 경우 v0.3.0까지는 VM에 대한 비용을 제공할 예정이지만, 향후 수집되는 모든 비용에 대한 배치처리 예정 중 
 -- 비용에 대한 출력은 모든 인스턴스에 대해서 할 예정이지만, 최적화에 대한 지원이 모든 범위가 가능한것은 아님

- 리소스 운용정보 취득 방법
 -- 인스턴스에 대한 고유 ID에 대한 정보가 필요, 취득방법에 대해서는 가이드 주시는 대로 개발 착수 예정 
 -- 태그로 매핑하는 경우도 마찬가지로, 태그에 포함된 인스턴스에 대한 값을 확인할 수 있으면 가능, tag정보에 대한 직접 수집은 어려움

- 최적화 정책에 따른 액션 수행 필요성 및 방안
 -- 현재까지 논의된 바가 없어, slack/mailing을 활용하는 방안으로 고려중
 -- action 방식에 대해 가이드 주시면 맞춰서 개발 예정
 -- 최적화의 경우 일반적으로 추천까지만 진행, 실제 최적화를 수행하는 것은 직접 수행하도록 권고함
 -- 예시로 이상비용에 대한 최적화를 고려할때, 테스트 및 배치성 인스턴스를 프로그램이 감지하여 바로 최적화를 진행한다면 문제가 발생할 수 있을것으로 예상

- API 문서 및 기능 제공 필요
 -- alarm기능 및 화면에서 비용을 출력하는 부분에 대한 api 문서 준비중

이상입니다.

**Reply by seokho-son** [reply]
@BSjaeyoung-kim 깔끔한 정리 감사합니다. 컨소시움 회의에서도 해당 내용으로 공유해주시면 되겠습니다. :)
